### PR TITLE
Reviving Linux support for certificate installation

### PR DESCRIPTION
**Pull request by ceuk**
Made the changes suggested in #156 and compared against the most recent commit in master.

### CODE REVIEW

**Comment by akrantz** [comment]
@asachin96 Could you take a look?

**Comment by asachin96** [comment]
The code looks good but can you place ca.crt file under office-addin-dev-certs folder in /usr/local/share/ca-certificates. This clears whats the purpose of the file, and doesn't conflict if there's already ca.crt file. And this needs to be created if not present. Eg. Install will be like
sudo mkdir -p  /usr/local/share/ca-certificates/office-addin-dev-certs && sudo cp ${caCertificatePath} /usr/local/share/ca-certificates/office-addin-dev-certs/ && sudo /usr/sbin/update-ca-certificates

**Comment by akrantz** [comment]
@TCourtneyOwen Please work to get the relevant changes made here to unblock doing add-in development on Linux.

File: packages/office-addin-dev-certs/src/install.ts [hunk 1]
```diff
@@ -15,6 +15,8 @@ function getInstallCommand(caCertificatePath: string, machine: boolean = false):
          return `powershell -ExecutionPolicy Bypass -File "${script}" ${machine ? "LocalMachine" : "CurrentUser"} "${caCertificatePath}"`;
       case "darwin": // macOS
          return `sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain '${caCertificatePath}'`;
+      case "linux":
+         return `sudo cp ${caCertificatePath} /usr/local/share/ca-certificates && sudo /usr/sbin/update-ca-certificates`;
```
Comment (VladimirMikulic) [inline packages/office-addin-dev-certs/src/install.ts]: It should be updated to `defaults.caCertificatePath`

**Comment by TheOpenDevProject** [comment]
@TCourtneyOwen Can we get this merged?

**Comment by TCourtneyOwen** [comment]
@TheOpenDevProject I will look into this today.  Frankly I am reluctant to merge this commit because Office Add-ins aren't supported on Linux in general.  By merging this change, we will support installing of dev-certs on Linux, but there will be many other aspects that aren't supported on Linux - I fear we may be opening ourselves up to many other issues that users encounter on Linux when they mistakenly assume that Linux is supported in general.

**Comment by VladimirMikulic** [comment]
@TCourtneyOwen the Linux community will help you fix them as well :)

**Comment by ceuk** [comment]
@TCourtneyOwen would it be beneficial to try and enumerate other potential issues Linux users might face rather than give up on them? Personally, I've used this branch to contribute to add-in development with no further issues.

Also, pragmatically speaking, given that this workaround is fairly trivial and publicly available, people will be developing on Linux regardless of whether this gets merged or not. 

You guys are free to reply to Linux-related issues with "we don't officially support Linux". Merging would just allow Linux users to continue to benefit from updates (and would stop this being raised every 6 months or so).

**Comment by TCourtneyOwen** [comment]
@ceuk It's not that we are resistant entirely to supporting Linux - it's that we haven't had the resources to do so yet.  Tell you what, on Monday I will get a CI Pipeline hooked up to this repo for Linux (currently we only have Windows and Mac) so that at least our tests run against Linux to ensure that this and any future changes don't cause issues.  If the CI loop run comes back clean (i.e. no failures), which I assume will be the case, we can merge this.  Does that sound reasonable?

**Comment by ceuk** [comment]
Sounds good to me!

**Comment by VladimirMikulic** [comment]
:tada: :tada: :tada:

**Comment by TCourtneyOwen** [comment]
@ceuk I added Linux to our pipeline, but the test phase fails on Linux because some of our other packages under Office-Addin-Scripts, which also get checked as part of the CI loop, explicitly throw exceptions if the process.platform is not Win32 or Mac.  I could potentially disable the test phase and just do install and build, both which currently pass on Linux, but this defeats the purpose on Linux.  The other alternative is to update the test code so that the failing tests only run on Win32 and Mac.  I am not entirely sure how many tests fall into this category however since the test phase bails out after the first failure

**Comment by akrantz** [comment]
I would like to add a couple of comments. First, I don't believe that we were able to validate yet that these changes for the dev-certs on Linux were working. There were still some comments that I don't believe were addressed.

Second, Linux is not currently a supported platform and we would need to do work as Courtney mentions to have it supported. 

We understand the desire to have this as a supported platform (for Web scenarios). I appreciate the work that has been done. I think it would be good for us to close on the changes needed and validate this for Linux so the dev-certs work there. It might take us longer to actually make the other changes required for supporting Linux as well.

I think the outstanding issues are:
1. the location where the .office-addin-dev-certs folder is created.
2. merging the latest changes from master into the branch. 
3. addressing unresolved comments.
4. validating that install, verify, uninstall all work properly on Linux.

**Comment by ceuk** [comment]
> I think the outstanding issues are:
> 
> 1. the location where the .office-addin-dev-certs folder is created.
> 2. merging the latest changes from master into the branch.
> 3. addressing unresolved comments.
> 4. validating that install, verify, uninstall all work properly on Linux.

@akrantz b95166d changes the default certificate installation path, I've rebased onto master and I've resolved Vladimir's comment about using `default.caCertificatePath` (c10f4d9) - so I think that should resolve 1, 2 and 3.

Regarding 4 and @TCourtneyOwen's comment. It seems like the monorepo architecture is going to make it difficult to validate. Perhaps we could specify which OS each package supports and only fail automatically skip/pass tests for package tests in unsupported OS's? 

It's a shame you can't record supported OS's in a package.json file as far as I'm aware. Perhaps a config file in the project root or something? 

I'm aware this is generating a fair bit of work to update tests/CI etc but can't think of a more expedient/pragmatic option unless anyone else has any ideas

**Comment by akrantz** [comment]
I'm going to see if we can easily skip the tests which are not applicable on Linux and close on this support. Thanks for making the updates and your understanding.

**Comment by TheOpenDevProject** [comment]
@TCourtneyOwen Thanks for the responsiveness, Microsoft has truly come along way to working with the FOSS and linux community, and a truly reasonable approach to integrating this branch. I'm with the others in that the linux community will work out its own bugs and push fixes.

**Comment by akrantz** [comment]
@ceuk I made a separate topic branch in my fork off of your branch. Usually you should use "git checkout -b <topic>" to make a separate branch for the pr, push that branch to your fork, and use that branch for the pull request. Because you just used "master" on your fork, I had to check out the branch with a different name. I therefore pushed the "dev-certs-linux" branch to my fork, and made another PR from that branch. 

See PR: https://github.com/OfficeDev/Office-Addin-Scripts/pull/274

**Comment by ceuk** [comment]
Hi @akrantz, apologies, when I first made this PR I thought it was just going to be a really quick patch and forking via GitHub is the quickest way. I'll endeavor to use a different branch name in the future.

I'm not sure if this is useful, but the name of my remote branch shouldn't matter too much IMO. You can checkout a local branch with any name you want and set it to track this remote/branch. 

e.g. in your local repo you could do:

`git remote add ceuk` 
`git fetch ceuk master` 
`git checkout -b dev-certs-linux ceuk/master` 
> Branch 'dev-certs-linux' set up to track remote branch 'master' from 'ceuk'.
> Switched to a new branch 'dev-certs-linux'

At which point you could just push any changes to ceuk/master which would appear in this PR:
`git push -f ceuk dev-certs-linux:master` 
or just
`git push -f ceuk HEAD:master`

I'll check out #247 though

**Comment by austinjp** [comment]
Hi. I wondered if any progress has been made on this?

I spent a while today rolling my own solution before discovering this PR. Just adding my voice to the Linux users who'd like this available. I use Office 365 over the web for work, but use Linux almost exclusively as my OS. I'd like to develop add-ins for Office 365, and it would be encouraging to know that the official tool manage dev certs on Linux without patching.

Thanks :)

**Comment by kubami** [comment]
I understand merging this is still going to take some time/decisions, as there was not a lot of movement on this or the other ( #156 ) PR regarding linux support.

In the meantime, I was unsuccessful in using this branch.
I have:
 1. generated a yeoman project: `yo office`,
 2. cloned this repo and checked out this branch
 3. `npm build` inside the `office-addin-dev-certs`
 4. `npm remove office-addin-dev-certs` + `npm install ./cloned-office-addins-repo/packages/office-addin-dev-certs` 
 
now the `npm run dev-server` passes the initial phase but spews errors like:
"TS2307: Cannot find module 'xregexp' or its corresponding type declarations."

the `npm run start:web`

```
> office-addin-taskpane@0.0.1 start:web
> office-addin-debugging start manifest.xml web

Debugging is being started...
App type: web
Debugging started.
```

**Comment by akrantz** [comment]
@kubami For the cloned office-addin-scripts repository, you'll want to do `npm install` and `npm run build` at the root of the repo. This will allow all of the packages to be built. To use the built office-addin-dev-certs package from the yo office project, go to that directory and to `npm link <path to office-addin-dev-certs>`, for example `npm link c:\git\office-addin-scripts\packages\office-addin-dev-certs`.  This will make the node_modules in the yo office project a link to the office-addin-dev-certs folder and it should use the files built from there.

**Comment by kubami** [comment]
@akrantz Thanks Adam, that helped. 

I can start the dev server with `npm run dev-server`. 
I am not sure if I do have to run `npm run start:web`, the output I get is:

```

> office-addin-taskpane@0.0.1 start:web
> office-addin-debugging start manifest.xml web

Debugging is being started...
App type: web
Debugging started.
```

For other that might be trying to run on linux. As mentioned in #156 , you need to go to: `https://localhost:3000` to accept the cert (note the http_s_)

The normal sideloading of the add-in works. 

Thank you!

**Comment by cortexcompiler** [comment]
What is holding this one up at this point?  Ironically I was trying to get an Outlook Add-in to work from Ubuntu on WSL2 and hit the same limitation.  I can revert to running the Add-In on Windows, but would be nice to keep my dev on linux.

**Comment by austinjp** [comment]
I'm having to use my own fork of this repo with a handful of very basic
hacks applied to get the certs working on my Linux laptop. It would really
be great if this issue could be resolved! 👍👌

**Comment by kubami** [comment]
Same here, me and my team are working with the cert patch I described above. 
It'd be great to have this merged in.

**Pull request closed without merging**